xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/XWikiServletURLFactory.java

```diff
@@ -607,12 +607,60 @@ public URL createAttachmentURL(String filename, String spaces, String name, Stri
                         querystring, xwikidb, context);
                 }
             } catch (XWikiException e) {
-                if (LOGGER.isErrorEnabled()) {
-                    LOGGER.error("Exception while trying to get attachment version !", e);
+                LOGGER.error("Exception while trying to get attachment version !", e);
+            }
+        }
+
+        // In the context of the current doc, no need to reload it.
+        if (isContextDoc(xwikidb, spaces, name, context)) {
+            XWikiAttachment attachment = context.getDoc().getAttachment(filename);
+            if (attachment != null) {
+                return createAttachmentRevisionURL(filename, spaces, name, attachment.getVersion(),
+                    querystring, xwikidb, context);
+            } else {
+                LOGGER.error("Exception while trying to get attachment [{}] from space [{}] and page [{}]!", filename,
+                    spaces, name);
+            }
+        } else {
+            String originalWikiId = context.getWikiId();
```

Comment:
Better play with the wiki reference instead of the id since that's what is stored in the XWikiContext.

Reply:
Fixed.